man/xml2gff.1
.Dd May 30, 2020
.Dt XML2GFF 1
.Os
.Sh NAME
.Nm xml2gff
.Nd XML to BioWare GFF converter
.Sh SYNOPSIS
.Nm xml2gff
.Ar input_file
.Ar output_file
.Sh DESCRIPTION
.Nm
converts XML files created by the
.Xr gff2xml 1
tool back into the BioWare GFF3 format.
.Pp
Unsupported Features:
.Bl -bullet -compact
.It
GFF4 (V4.0/V4.1)
.It
Color codes in strings
.El
.Pp
The format of the input XML is rather simple, but there are many different elements.
.Bd -literal
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<gff3 type="ID">
  <struct id="4294967295">
    <list label="test_list">
      <struct label="list_struct" id="0">
        ...
      </struct>
    </list>
    <struct label="nonlist_struct" id="0">
      ...
    </struct>
    <byte label="value1">1</byte>
    <char label="value2">A</char>
    <sint16 label="value3">-1</sint16>
    <sint32 label="value4">-1</sint32>
    <sint64 label="value5">-1</sint64>
    <uint16 label="value6">1</uint16>
    <uint32 label="value7">1</uint32>
    <uint64 label="value8">1</uint64>
    <float label="value10">1.0</float>
    <double label="value11">1.0</double>
    <exostring label="value12">Test String</exostring>
    <resref label="value13">test_resource</resref>
    <locstring label="value14" strref="46">
      <string language="0">Localized String</string>
      ...
    </locstring>
    <void label="value15">...</void>
    <orientation label="value16">
      <double>1.0<double>
      <double>1.0<double>
      <double>1.0<double>
      <double>1.0<double>
    </orientation>
    <vector label="value17">
      <double>1.0<double>
      <double>1.0<double>
      <double>1.0<double>
    </vector>
    <strref label="value13" strref="12">
  </struct>
</gff3>
.Ed
.Pp
The root element is either
.Dq gff3
or
.Dq gff4
which determines the GFF format to write (GFF4 is currently unsupported)
and it must have a type property with a maximum of 4 characters. If there
are more than 4, only the first 4 will be used.
.Pp
The element under the gff tag always needs to be a
.Dq struct
tag with the id
.Dq 4294967295
after that all other tags which are childs of the root struct in the sample above are possible
to use.
.Pp
The tags
.Dq vector
and
.Dq orientation
need to have 3 and 4 child
.Dq double
tags.
.Dq locstring
tags can have multiple
.Dq string
tags with a language property which combines language and gender.
struct tags including the root one can only have
.Dq byte ,
.Dq char ,
.Dq sint16 ,
.Dq sint32 ,
.Dq sint64 ,
.Dq uint16 ,
.Dq uint32 ,
.Dq uint64 ,
.Dq float ,
.Dq double ,
.Dq exostring ,
.Dq resref ,
.Dq locstring ,
.Dq data ,
.Dq orientation ,
.Dq vector ,
.Dq strref ,
.Dq list
and
.Dq struct
tags.
The
.Dq list
accepts only struct child tags.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.El
.Pp
.Bl -tag -width xxxx -compact
.It Fl Fl nwn
Write LocStrings in an encoding appropriate for
.Em Neverwinter Nights .
.It Fl Fl nwn2
Write LocStrings in an encoding appropriate for
.Em Neverwinter Nights 2 .
.It Fl Fl kotor
Write LocStrings in an encoding appropriate for
.Em Knights of the Old Republic .
.It Fl Fl kotor2
Write LocStrings in an encoding appropriate for
.Em Knights of the Old Republic II .
.It Fl Fl jade
Write LocStrings in an encoding appropriate for
.Em Jade Empire .
.It Fl Fl witcher
Write LocStrings in an encoding appropriate for
.Em The Witcher .
.It Fl Fl dragonage
Write LocStrings in an encoding appropriate for
.Em Dragon Age: Origins .
.It Fl Fl dragonage2
Write LocStrings in an encoding appropriate for
.Em Dragon Age II .
.It Fl Fl encoding Ar str
Override an encoding.
The string has to be of the form n=encoding, for example 0=cp-1252
to override the encoding of the (ungendered) language ID 0 to be
Windows codepage 1252.
To override several encodings, specify the --encoding parameter
multiple times.
.El
.Pp
.Bl -tag -width xxxx -compact
.It Fl Fl v32
Create a GFF3 V3.2. This is the default for all games except
.Em The Witcher .
.It Fl Fl v33
Create a GFF3 V3.3. This is the default for
.Em The Witcher .
.El
.Pp
.Bl -tag -width xxxx -compact
.It Ar input_file
The XML file to convert.
If the input is -, the XML data is read from
.Dv stdin .
.It Ar output_file
The GFF file will be written there.
.El
.Sh EXAMPLES
Convert
.Pa file1.xml
into a V3.2 GFF file:
.Pp
.Dl $ xml2gff file1.xml file2.gff
.Pp
Convert stdin
into a V3.2 GFF file:
.Pp
.Dl $ xml2gff - file1.gff
.Pp
Convert
.Pa file1.xml
into a V3.2 GFF file, using
.Em Neverwinter Nights
encoding tables:
.Pp
.Dl $ xml2gff file1.xml file2.gff
.Pp
Convert
.Pa file1.xml
into a V3.2 GFF file, and encode language ID 0 in LocStrings
as Windows CP-1250:
.Dl $ gff2xml --encoding 0=cp1250 file1.xml file2.gff
.Sh SEE ALSO
.Xr gff2xml 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/xml2gff.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert XML files back into GFF.
 */

#include "src/version/version.h"

#include "src/common/scopedptr.h"
#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdinstream.h"
#include "src/common/encoding.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"
#include "src/aurora/language.h"

#include "src/xml/gffcreator.h"

#include "src/util.h"

typedef std::map<uint32, Common::Encoding> EncodingOverrides;

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Aurora::GameID &game, EncodingOverrides &encOverrides,
                      XML::GFFCreator::GFF3Version &gff3Version);

bool parseEncodingOverride(const Common::UString &arg, EncodingOverrides &encOverrides);

void createGFF(const Common::UString &inFile, const Common::UString &outFile, XML::GFFCreator::GFF3Version gff3Version);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Aurora::GameID game = Aurora::kGameIDUnknown;
		EncodingOverrides encOverrides;
		XML::GFFCreator::GFF3Version gff3Version = XML::GFFCreator::GFF3Version::Unknown;

		int returnValue = 1;
		Common::UString inFile, outFile;

		if (!parseCommandLine(args, returnValue, inFile, outFile, game, encOverrides, gff3Version))
			return returnValue;

		LangMan.declareLanguages(game);

		for (EncodingOverrides::const_iterator e = encOverrides.begin(); e != encOverrides.end(); ++e)
			LangMan.overrideEncoding(e->first, e->second);

		if (gff3Version == XML::GFFCreator::GFF3Version::Unknown) {
			if (game == Aurora::kGameIDWitcher)
				gff3Version = XML::GFFCreator::GFF3Version::V3_3;
			else
				gff3Version = XML::GFFCreator::GFF3Version::V3_2;
		}

		createGFF(inFile, outFile, gff3Version);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseEncodingOverride(const Common::UString &arg, EncodingOverrides &encOverrides) {
	Common::UString::iterator sep = arg.findFirst('=');
	if (sep == arg.end())
		return false;

	uint32 id = 0xFFFFFFFF;
	try {
		Common::parseString(arg.substr(arg.begin(), sep), id);
	} catch (...) {
	}

	if (id == 0xFFFFFFFF)
		return false;

	Common::Encoding encoding = Common::parseEncoding(arg.substr(++sep, arg.end()));
	if (encoding == Common::kEncodingInvalid) {
		status("Unknown encoding \"%s\"", arg.substr(sep, arg.end()).c_str());
		return false;
	}

	encOverrides[id] = encoding;
	return true;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Aurora::GameID &game, EncodingOverrides &encOverrides,
                      XML::GFFCreator::GFF3Version &gff3Version) {

	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::Callback;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Aurora::GameID;
	using XML::GFFCreator;

	std::vector<Common::UString> args;
	NoOption inFileOpt(false, new ValGetter<Common::UString &>(inFile, "input file"));
	NoOption outFileOpt(false, new ValGetter<Common::UString &>(outFile, "output file"));
	Parser parser(argv[0], "XML to BioWare GFF converter",
	              "If the input file is -, the input is read from stdin.\n\n"
	              "The XML root tag determines, if a GFF3 or GFF4 file will be written\n"
	              "and the type property determines which GFF ID will be written. GFF IDs\n"
	              "can be at most 4 characters long.\n\n"
	              "Depending on the game, LocStrings in GFF files might be encoded in various\n"
	              "ways and there's no way to autodetect how. If a game is specified, the\n"
	              "encoding tables for this game are used. Otherwise, xml2gff writes plain\n"
	              "ASCII strings, removing other characters.\n\n"
	              "Additionally, the --encoding parameter can be used to override the encoding\n"
	              "for a specific language ID. The string has to be of the form n=encoding,\n"
	              "for example 0=cp-1252 to override the encoding of the (ungendered) language\n"
	              "ID 0 to be Windows codepage 1252. To override several encodings, specify\n"
	              "the --encoding parameter multiple times.\n",
	              returnValue,
	              makeEndArgs(&inFileOpt, &outFileOpt));

	parser.addSpace();
	parser.addOption("nwn", "Use Neverwinter Nights encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN, game)));
	parser.addOption("nwn2", "Use Neverwinter Nights 2 encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("kotor", "Use Knights of the Old Republic encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR, game)));
	parser.addOption("kotor2", "Use Knights of the Old Republic II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR2, game)));
	parser.addOption("jade", "Use Jade Empire encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
	parser.addOption("witcher", "Use The Witcher encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDWitcher, game)));
	parser.addOption("dragonage", "Use Dragon Age encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge, game)));
	parser.addOption("dragonage2", "Use Dragon Age II encodings", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge2, game)));
	parser.addSpace();
	parser.addOption("encoding", "Override an encoding", kContinueParsing,
	                 new Callback<EncodingOverrides &>("str", parseEncodingOverride, encOverrides));
	parser.addSpace();
	parser.addOption("v32", "Create GFF3 V3.2 file (default)", kContinueParsing,
	                 makeAssigners(new ValAssigner<GFFCreator::GFF3Version>(GFFCreator::GFF3Version::V3_2, gff3Version)));
	parser.addOption("v33", "Create GFF3 V3.3 file (default for The Witcher)", kContinueParsing,
	                 makeAssigners(new ValAssigner<GFFCreator::GFF3Version>(GFFCreator::GFF3Version::V3_3, gff3Version)));

	return parser.process(argv);
}

void createGFF(const Common::UString &inFile, const Common::UString &outFile, XML::GFFCreator::GFF3Version gff3Version) {
	Common::ScopedPtr<Common::ReadStream> xml(openFileOrStdIn(inFile));
	Common::ScopedPtr<Common::WriteStream> gff(openFileOrStdOut(outFile));

	XML::GFFCreator::create(*gff, *xml, inFile, gff3Version);

	gff->flush();
}
